ZENKO-1730 lifecycle checks data mover progress

jonathan-gramain opened the pull request:
**Summary**

To prevent lifecycle from running a batch again when data mover still
has some messages to process from the previous batch, introduce a new
topic offset snapshot mechanism to check progress of the data mover,
in addition to the existing bucket and object tasks consumer lag
checks.

Also refactored how checks are done: all the mechanisms to check
offsets have been moved to a common class KafkaBacklogMetrics for
clarity, that is shared between all components either publishing or
checking the offsets.

**Details**

Checking the consumer lag for the data mover topic, like for bucket
and object tasks would not be ideal, as the data mover might be used
by other processes which might introduce unrelated consumer lag.

Instead, we use a snapshot/check system, where:

- the bucket processor takes a snapshot of the data mover topic
  offsets after it has finished processing a bucket

- the conductor checks whether the latest data mover consumer offsets
  are beyond the latest snapshot offset values, which means that what
  the bucket processor published to the data mover has already been
  fully processed, but regardless of messages coming from other
  services. It also still checks the bucket processor and object
  processor consumer lag as before.

jonathan-gramain committed:
ZENKO-1730 lifecycle checks data mover progress

**Summary**

To prevent lifecycle from running a batch again when data mover still
has some messages to process from the previous batch, introduce a new
topic offset snapshot mechanism to check progress of the data mover,
in addition to the existing bucket and object tasks consumer lag
checks.

Also refactored how checks are done: all the mechanisms to check
offsets have been moved to a common class KafkaBacklogMetrics for
clarity, that is shared between all components either publishing or
checking the offsets.

**Details**

Checking the consumer lag for the data mover topic, like for bucket
and object tasks would not be ideal, as the data mover might be used
by other processes which might introduce unrelated consumer lag.

Instead, we use a snapshot/check system, where:

- the bucket processor takes a snapshot of the data mover topic
  offsets after it has finished processing a bucket

- the conductor checks whether the latest data mover consumer offsets
  are beyond the latest snapshot offset values, which means that what
  the bucket processor published to the data mover has already been
  fully processed, but regardless of messages coming from other
  services. It also still checks the bucket processor and object
  processor consumer lag as before.

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @bennettbuchanan

* @vrancurel

* @jonathan-gramain

bennettbuchanan commented on lib/KafkaBacklogMetrics.js:
```diff
@@ -0,0 +1,467 @@
+const zookeeperHelper = require('./clients/zookeeper');
+
+const CheckConditionError = {};
+const NoNodeError = {};
```
Why not make these instances of `Error`?

jonathan-gramain replied:
I could, but I don't know if that would bring any value, but I'm open to changing this if it does! I though since those errors never leave the inner class functions, they can be simple empty objects.

bennettbuchanan replied:
Definitely it's functionally the same. In my mind I was expecting the objects to be populated with values in some way, so by making them an `Error` I would know they'd be used as such. It's a minor readability point though.

jonathan-gramain replied:
done

philipyoo commented on tests/functional/lifecycle/LifecycleTask.js:
```diff
@@ -344,6 +344,12 @@ class ProducerMock {
     }
 }
 
+class KafkaBacklogMetricsMock {
```
looks like ProducerMock on L#305 just needs to add a mock of method `getKafkaProducer` to get these functional tests working again

jonathan-gramain commented:
@philipyoo @bennettbuchanan I added a new commit to fix and test better the snapshot mechanism as there was a couple bugs there.

I also added another commit to fix a typo in the BackbeatMetadataProxy test mock of lifecycle tests (headObject -> headLocation), as I saw the tests failing when trying locally, but the tests are passing on the CI with the typo, that looks surprising to me and I don't have an explanation so far.

philipyoo commented on lib/KafkaBacklogMetrics.js:
```diff
@@ -0,0 +1,476 @@
+    }
+
+    /**
+     * Check whether the given consumer group lags beyond the maximum
```
minor: rather than max lag, a check on consumer info

jonathan-gramain replied:
Good catch, a cut&paste typo

jonathan-gramain commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/7.4`
* `development/7.5`
* `development/8.0`
* `stabilization/7.4.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/7.4`
* `development/7.5`
* `development/8.0`
* `stabilization/7.4.3`



Please check the status of the associated issue ZENKO-1730.

Goodbye jonathan-gramain.
